File: test/test-imports/java/expected-from-config/k8s/src/main/java/k8s/ServiceAccount.java
package k8s;

/**
 * ServiceAccount binds together: * a name, understood by users, and perhaps by peripheral systems, for an identity * a principal that can be authenticated and authorized * a set of secrets.
 */
@javax.annotation.Generated(value = "jsii-pacmak/1.1.0 (build df55f5e)", date = "2020-05-13T15:58:47.425Z")
@software.amazon.jsii.Jsii(module = k8s.$Module.class, fqn = "k8s.ServiceAccount")
public class ServiceAccount extends org.cdk8s.ApiObject {

    protected ServiceAccount(final software.amazon.jsii.JsiiObjectRef objRef) {
        super(objRef);
    }

    protected ServiceAccount(final software.amazon.jsii.JsiiObject.InitializationMode initializationMode) {
        super(initializationMode);
    }

    /**
     * Defines a "io.k8s.api.core.v1.ServiceAccount" API object.
     * <p>
     * @param scope the scope in which to define this object. This parameter is required.
     * @param name a scope-local name for the object. This parameter is required.
     * @param options configuration options.
     */
    public ServiceAccount(final @org.jetbrains.annotations.NotNull software.constructs.Construct scope, final @org.jetbrains.annotations.NotNull java.lang.String name, final @org.jetbrains.annotations.Nullable k8s.ServiceAccountOptions options) {
        super(software.amazon.jsii.JsiiObject.InitializationMode.JSII);
        software.amazon.jsii.JsiiEngine.getInstance().createNewObject(this, new Object[] { java.util.Objects.requireNonNull(scope, "scope is required"), java.util.Objects.requireNonNull(name, "name is required"), options });
    }

    /**
     * Defines a "io.k8s.api.core.v1.ServiceAccount" API object.
     * <p>
     * @param scope the scope in which to define this object. This parameter is required.
     * @param name a scope-local name for the object. This parameter is required.
     */
    public ServiceAccount(final @org.jetbrains.annotations.NotNull software.constructs.Construct scope, final @org.jetbrains.annotations.NotNull java.lang.String name) {
        super(software.amazon.jsii.JsiiObject.InitializationMode.JSII);
        software.amazon.jsii.JsiiEngine.getInstance().createNewObject(this, new Object[] { java.util.Objects.requireNonNull(scope, "scope is required"), java.util.Objects.requireNonNull(name, "name is required") });
    }

    /**
     * A fluent builder for {@link k8s.ServiceAccount}.
     */
    public static final class Builder {
        /**
         * @return a new instance of {@link Builder}.
         * @param scope the scope in which to define this object. This parameter is required.
         * @param name a scope-local name for the object. This parameter is required.
         */
        public static Builder create(final software.constructs.Construct scope, final java.lang.String name) {
            return new Builder(scope, name);
        }

        private final software.constructs.Construct scope;
        private final java.lang.String name;
        private k8s.ServiceAccountOptions.Builder options;

        private Builder(final software.constructs.Construct scope, final java.lang.String name) {
            this.scope = scope;
            this.name = name;
        }

        /**
         * AutomountServiceAccountToken indicates whether pods running as this service account should have an API token automatically mounted.
         * <p>
         * Can be overridden at the pod level.
         * <p>
         * @return {@code this}
         * @param automountServiceAccountToken AutomountServiceAccountToken indicates whether pods running as this service account should have an API token automatically mounted. This parameter is required.
         */
        public Builder automountServiceAccountToken(final java.lang.Boolean automountServiceAccountToken) {
            this.options().automountServiceAccountToken(automountServiceAccountToken);
            return this;
        }

        /**
         * ImagePullSecrets is a list of references to secrets in the same namespace to use for pulling any images in pods that reference this ServiceAccount.
         * <p>
         * ImagePullSecrets are distinct from Secrets because Secrets can be mounted in the pod, but ImagePullSecrets are only accessed by the kubelet. More info: https://kubernetes.io/docs/concepts/containers/images/#specifying-imagepullsecrets-on-a-pod
         * <p>
         * @return {@code this}
         * @param imagePullSecrets ImagePullSecrets is a list of references to secrets in the same namespace to use for pulling any images in pods that reference this ServiceAccount. This parameter is required.
         */
        public Builder imagePullSecrets(final java.util.List<k8s.LocalObjectReference> imagePullSecrets) {
            this.options().imagePullSecrets(imagePullSecrets);
            return this;
        }

        /**
         * Standard object's metadata.
         * <p>
         * More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
         * <p>
         * @return {@code this}
         * @param metadata Standard object's metadata. This parameter is required.
         */
        public Builder metadata(final k8s.ObjectMeta metadata) {
            this.options().metadata(metadata);
            return this;
        }

        /**
         * Secrets is the list of secrets allowed to be used by pods running using this ServiceAccount.
         * <p>
         * More info: https://kubernetes.io/docs/concepts/configuration/secret
         * <p>
         * @return {@code this}
         * @param secrets Secrets is the list of secrets allowed to be used by pods running using this ServiceAccount. This parameter is required.
         */
        public Builder secrets(final java.util.List<k8s.ObjectReference> secrets) {
            this.options().secrets(secrets);
            return this;
        }

        /**
         * @returns a newly built instance of {@link k8s.ServiceAccount}.
         */
        public k8s.ServiceAccount build() {
            return new k8s.ServiceAccount(
                this.scope,
                this.name,
                this.options != null ? this.options.build() : null
            );
        }

        private k8s.ServiceAccountOptions.Builder options() {
            if (this.options == null) {
                this.options = new k8s.ServiceAccountOptions.Builder();
            }
            return this.options;
        }
    }
}

File: packages/cdk8s-plus/src/service-account.ts
import { ApiObject, Lazy } from 'cdk8s';
import { Construct } from 'constructs';
import { Resource, ResourceProps, IResource } from './base';
import * as k8s from './imports/k8s';
import { ISecret } from './secret';
import { undefinedIfEmpty } from './utils';


/**
 * Properties for initialization of `ServiceAccount`.
 */
export interface ServiceAccountProps extends ResourceProps {

}

export interface IServiceAccount extends IResource {

}

/**
 * Properties for initialization of `ServiceAccount`.
 */
export interface ServiceAccountProps {
  /**
   * List of secrets allowed to be used by pods running using this
   * ServiceAccount.
   *
   * @see https://kubernetes.io/docs/concepts/configuration/secret
   */
  readonly secrets?: ISecret[];
}

/**
 * A service account provides an identity for processes that run in a Pod.
 *
 * When you (a human) access the cluster (for example, using kubectl), you are
 * authenticated by the apiserver as a particular User Account (currently this
 * is usually admin, unless your cluster administrator has customized your
 * cluster). Processes in containers inside pods can also contact the apiserver.
 * When they do, they are authenticated as a particular Service Account (for
 * example, default).
 *
 * @see https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account
 */
export class ServiceAccount extends Resource implements IServiceAccount {

  /**
   * Imports a service account from the cluster as a reference.
   * @param name The name of the service account resource.
   */
  public static fromServiceAccountName(name: string): IServiceAccount {
    return { name: name };
  }

  /**
   * @see base.Resource.apiObject
   */
  protected readonly apiObject: ApiObject;

  private readonly _secrets: ISecret[];

  constructor(scope: Construct, id: string, props: ServiceAccountProps = { }) {
    super(scope, id, { metadata: props.metadata });

    this._secrets = props.secrets ?? [];

    this.apiObject = new k8s.KubeServiceAccount(this, 'Resource', {
      metadata: props.metadata,
      secrets: Lazy.any({ produce: () => undefinedIfEmpty(this._secrets.map(s => ({ name: s.name }))) }),
    });
  }

  /**
   * Allow a secret to be accessed by pods using this service account.
   * @param secret The secret
   */
  public addSecret(secret: ISecret) {
    this._secrets.push(secret);
  }

  /**
   * List of secrets allowed to be used by pods running using this service
   * account.
   *
   * Returns a copy. To add a secret, use `addSecret()`.
   */
  public get secrets() {
    return [...this._secrets];
  }
}